### PR TITLE
Anomaly Detection Model on Time Series data

### LINKED ISSUE #6074: [Machine Learning] Anomaly Detection Model on Time Series data
#### Proposed title of article
[Machine Learning] Anomaly Detection Model on Time Series data

#### Proposed article introduction
Time-series data is a sequence of data points that occur in successive order over some period of time.  The time can be in years, days, hours, or minutes. Models then use this dataset to identify patterns and make future predictions. But this dataset may have anomalies,  anomalies are observations or data points that deviate from a dataset’s normal behavior. Anomalous data can indicate critical incidents or even potential opportunities. 

Anomalies in the dataset have a negative impact on the model's performance. They give wrong prediction results. Anomaly detection is used to identify data points that deviate from a dataset's normal behavior so that we have a uniform dataset. 

In this tutorial, we will use Facebook Prophet for anomaly detection. Facebook Prophet is a good library used for building time series modeling. After removing the anomalies, we will use the clean dataset to build our machine learning model.

#### Key takeaways 
1. Loading the time series dataset.
2. Anamoly detection techniques using Facebook Prophet.
3. Removing the anomalies from our dataset.
4. Using the clean dataset to build a time series model.

#### Article quality
In this tutorial we will discuss the basics of time series in detail, this enables a reader to fully understand this type of machine learning before we introduce anomaly detection. I will also ensure that the reader knows the various libraries that can be used for anomaly detection. Finally, we will implement Facebook Prophet to detect and remove the anomaly dataset, the clean dataset will then used to build the time series model.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by charlesndirutu33**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #6074

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
Plagiarism check done ✅

File: content/articles/anomaly-detection-model-on-time-series-data/index.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+- Have some knowledge on [time series](/engineering-education/introduction-to-time-series/)
+- Understand [time series decomposition in python](/engineering-education/time-series-decomposition-in-python/)
+- Know how to build a [time series model](/engineering-education/building-a-time-series-weather-forecasting-application-in-python/)
+- Use [Google Colab notebook](https://research.google.com/)
+
+### Time series dataset
+The dataset used in this tutorial is the New York Taxi dataset. The dataset was recorded on a half-hour basis for 6 months. The dataset shows the number of active taxi drivers in New York city every half-hour. 
+
+We will use this dataset to build a times series model using Facebook Prophet. The model will be used to predict the number of active taxi drivers in New York.
```
Comment (wilkistermumbi) [inline content/articles/anomaly-detection-model-on-time-series-data/index.md:54]:
```suggestion
We will use this dataset to build a times series model using Facebook Prophet. We will use the model to predict the number of active taxi drivers in New York. 
```

Comment (wilkistermumbi) [reply content/articles/anomaly-detection-model-on-time-series-data/index.md:54]: Always use active voice

File: content/articles/anomaly-detection-model-on-time-series-data/index.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+)
+fig.show()
+```
+From the code above, We have initialized the x-axis as `timestamp`. The y-axis is the `value`. And the title of the line chart is `NYC Taxi Demand`. We have added labels that can show the data points within 1 year, 3 years, or 5 years. 
+
+The diagram is shown below:
+
+![Line chart](/engineering-education/anomaly-detection-model-on-time-series-data/line-chart.jpg)
+
+The image above shows the number of active taxis with a period of July 2014 to Jan 2015.
```
Comment (wilkistermumbi) [inline content/articles/anomaly-detection-model-on-time-series-data/index.md:155]:
```suggestion
The image above shows the number of active taxis within the period of July 2014 to Jan 2015. 
```

File: content/articles/anomaly-detection-model-on-time-series-data/index.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+Let's initialize our model.
+
+### Initializing the model
+We will initialize our model using the following code:
+
+```python
+model = Prophet(changepoint_range=0.95)
+```
+We use a `changepoint_range=0.95` to increase the confidence interval of our output. The confidence interval determines if a prediction is right or wrong. The higher confidence interval increases the chances of the model making accurate predictions
```
Comment (wilkistermumbi) [inline content/articles/anomaly-detection-model-on-time-series-data/index.md:204]:
```suggestion
We use a `changepoint_range=0.95` to increase the confidence interval of our output. The confidence interval determines if a prediction is right or wrong. The higher confidence interval increases the chances of the model making accurate predictions.
```

File: content/articles/anomaly-detection-model-on-time-series-data/index.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+Let's fit the model to our train dataset.
+
+### Model fitting
+We fit the model to the training dataset so that the model can learn from it. 
+
+```python
+model.fit(train_dataset)
+```
+This will train the model and the model will be ready to make future predictions.
```
Comment (wilkistermumbi) [inline content/articles/anomaly-detection-model-on-time-series-data/index.md:216]:
```suggestion
This code will train the model. After training, the model will now be ready to make future predictions.
```

File: content/articles/anomaly-detection-model-on-time-series-data/index.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+### Making predictions using the test dataset
+The test dataset contains timestamp values past `2015-01-27`. The test dataset has 120 hours. We are forecasting for the 120 hours in the test dataset.
+
+```python
+future = model.make_future_dataframe(periods=120, freq='H')
+future.tail()
+```
+From the code above `periods=120` shows the 120 hours recorded in the test dataset. `freq='H'` will give the prediction values in hourly-interval. `future.tail()` will show the five last rows of the `ds` column. This is the last five hours the model will predict.
```
Comment (wilkistermumbi) [inline content/articles/anomaly-detection-model-on-time-series-data/index.md:225]:
```suggestion
From the code above, `periods=120` shows the 120 hours recorded in the test dataset. `freq='H'` will give the prediction values in hourly-interval. `future.tail()` will show the five last rows of the `ds` column. It shows the last five hours the model will predict.
```

File: content/articles/anomaly-detection-model-on-time-series-data/index.md [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+ dict(count=1, label='1y', step='year', stepmode='backward'),
+ dict(count=2, label='3y', step='year', stepmode='backward'),
+ dict(count=3, label='5y', step='year', stepmode='backward'),
+ dict(step='all')
+        ])
+    )
+)
+fig.show()
+```
+From the code above we have set `ds` as the x-axis and `y` as the y-axis. The title of the scatter plot is `NYC Taxi Demand`. Also, the anomalies detected will be given a different color. When the code is executed, it output the following diagram:
```
Comment (wilkistermumbi) [inline content/articles/anomaly-detection-model-on-time-series-data/index.md:322]:
```suggestion
From the code above, we have set `ds` as the x-axis and `y` as the y-axis. The title of the scatter plot is `NYC Taxi Demand`. Also, the anomalies detected will be given a different color. When the code is executed, it outputs the following diagram:
```

File: content/articles/anomaly-detection-model-on-time-series-data/index.md [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+
+![Scatter plot](/engineering-education/anomaly-detection-model-on-time-series-data/scatter-plot.jpg)
+
+In the image above, the detected anomalies have a red color. The normal data points have a blue color. The scatter plot has helped us to detect the outliers available in our dataset.
+
+### Conclusion
+In this tutorial, we have learned anomaly detection on time series data. We used the New York Taxi dataset to train our model. Using the Facebook Prophet library, we were able to build a time series model. The model was used to predict the number of active taxi drivers in New York.
+
+We then used a scatter plot to detect anomalies in our dataset. Using this tutorial, a reader should be able to detect anomalies on time series data.
```
Comment (wilkistermumbi) [inline content/articles/anomaly-detection-model-on-time-series-data/index.md:331]:
```suggestion
In this tutorial, we have learned anomaly detection on time series data. We used the New York Taxi dataset to train our model. Using the Facebook Prophet library, we built a time series model. The model was used to predict the number of active taxi drivers in New York.

We then used a scatter plot to detect anomalies in our dataset. Using this tutorial, a reader should detect anomalies on time series data.
```

**Comment by wilkistermumbi** [comment]
@charlesndirutu33 Please work on the revisions before we can move on for tech review.

**Comment by charlesndirutu33** [comment]
Hello @wilkistermumbi, I have made the requested changes
Thanks

**Comment by hectorkambow** [comment]
@wilkistermumbi  

If it looks like any more edits are needed - we will be closing this PR.

**Comment by hectorkambow** [comment]
@wilkistermumbi any update on this?

**Comment by wilkistermumbi** [comment]
@charlesndirutu33 Besides the minor commits, this is a very educative tutorial. It will benefit our readers. You've paid attention to explaining minor details that authors tend to ignore, even your hero image tells it all. You put a lot of thought into it. Keep it up.

**Comment by hectorkambow** [comment]
Jumping on this one

**Comment by hectorkambow** [comment]
nicely done - some minor edits on my end
